File: src/main/kotlin/com/mrgrd56/app/utils/TextAnalyzer.java
package com.mrgrd56.app.utils;

import java.util.ArrayList;
import java.util.Map;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

public class TextAnalyzer {
    private static final Pattern WORDS_REGEX = Pattern.compile("\\b[\\w-'’]+\\b");

    public static Map<String, Integer> countWords(String text) {
        var matcher = WORDS_REGEX.matcher(text);

        var matches = new ArrayList<String>();

        while (matcher.find()) {
            matches.add(matcher.group());
        }

        return matches.stream()
                .collect(Collectors.groupingBy(match -> match),
                        Collectors.mapping(group -> group));
    }
}

File: src/main/kotlin/com/mrgrd56/app/utils/TextAnalyzer.kt
package com.mrgrd56.app

val WORDS_REGEX = Regex("\\b[\\w-'’]+\\b")

fun countWords(text: String): Map<String, Int> {
    return WORDS_REGEX.findAll(text)
        .groupBy { it.value }
        .mapValues { it.value.size }
        .toList()
        .sortedByDescending { (_, v) -> v }
        .toMap()
}